crates/contract/src/call.rs

```diff
@@ -0,0 +1,404 @@
+    /// `max_fee_per_gas` and `max_priority_fee_per_gas` to the same value
+    pub fn gas_price(mut self, gas_price: U256) -> Self {
+        self.request = self.request.gas_price(gas_price);
+        self
+    }
+
+    /// Sets the `value` field in the transaction to the provided value
+    pub fn value(mut self, value: U256) -> Self {
+        self.request = self.request.value(value);
+        self
+    }
+
+    /// Sets the `nonce` field in the transaction to the provided value
+    pub fn nonce(mut self, nonce: U64) -> Self {
+        self.request = self.request.nonce(nonce);
+        self
+    }
+
+    /// Sets the `block` field for sending the tx to the chain
+    pub const fn block(mut self, block: BlockId) -> Self {
+        self.block = Some(block);
+        self
+    }
+
+    /// Sets the [state override set](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set).
+    ///
+    /// # Note
+    ///
+    /// Not all client implementations will support this as a parameter to `eth_call`.
+    pub fn state(mut self, state: StateOverride) -> Self {
+        self.state = Some(state);
+        self
+    }
+
+    /// Returns the underlying transaction's ABI-encoded data.
+    pub fn calldata(&self) -> &Bytes {
+        self.request.input.input().expect("set in the constructor")
+    }
+
+    /// Returns the estimated gas cost for the underlying transaction to be executed
+    pub async fn estimate_gas(&self) -> Result<U256> {
+        self.provider.estimate_gas(self.request.clone(), self.block).await.map_err(Into::into)
+    }
+
+    /// Queries the blockchain via an `eth_call` without submitting a transaction to the network.
+    ///
+    /// Returns the decoded the output by using the provided decoder.
+    /// If this is not desired, use [`call_raw`](Self::call_raw) to get the raw output data.
+    pub async fn call(&self) -> Result<D::Output> {
+        let data = self.call_raw().await?;
+        self.decode_output(data, false)
+    }
+
+    /// Queries the blockchain via an `eth_call` without submitting a transaction to the network.
+    ///
+    /// Does not decode the output of the call, returning the raw output data instead.
+    ///
+    /// See [`call`](Self::call) for more information.
+    pub async fn call_raw(&self) -> Result<Bytes> {
+        if let Some(state) = &self.state {
+            self.provider.call_with_overrides(self.request.clone(), self.block, state.clone()).await
+        } else {
+            self.provider.call(self.request.clone(), self.block).await
+        }
+        .map_err(Into::into)
+    }
+
+    /// Decodes the output of a contract function using the provided decoder.
+    #[inline]
+    pub fn decode_output(&self, data: Bytes, validate: bool) -> Result<D::Output> {
+        self.decoder.abi_decode_output(data, validate)
+    }
+
+    /// Signs and broadcasts this call as a transaction.
+    pub async fn send(&self) -> Result<()> {
+        todo!()
+    }
```

Comment:
Perhaps we can move this behind a Provider trait bound, requiring that the provider supports signing

Reply:
we want it to work with eth_sendTransaction too, so i think just need to modify the comment